### PR TITLE
Add BLAKE2b key types

### LINKED ISSUE #1800: Add BLAKE2b ID hash functions
- New key types which use BLAKE2b-256 for ID hash (and MAC?).
- `authenticated`: no encryption
- `repokey|keyfile-blake2`: the usual encryption.

From #1044 

---

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/38938088-add-blake2b-id-hash-functions)

### PR DESCRIPTION
**Pull request by enkore**
The BLAKE2b implementation is triple licensed under the CC0, OpenSSL and Apache 2.0 licenses, all of which are compatible with our BSD 3 clause.

The BLAKE2b implementation files are copied verbatim from git sha1 https://github.com/BLAKE2/BLAKE2/commit/d4418edab4b9e7595973f876f82d9fecf54b9d19 , completely unmodified to make maintenance as easy as possible.

This actually adds very little new code, since almost everything is reused. Note that the encrypt/decrypt code in AuthenticatedKey is copied from PlaintextKey (enc_key, mac_key are unused but still there, no harm done).

#1034 does not apply perfectly cleanly, but all merge conflicts are the simple kind (import lines in two files, one hunk in the crypto tests where git makes up a conflict where both branches added separate tests, the selftest count).

The error message from old borg looks like this:

```
Unsupported payload type 5. A newer version is required to access this repository.
```

Fixes #1800

### CODE REVIEW

File: src/borg/testsuite/key.py [hunk 1]
```diff
@@ -193,6 +229,13 @@ def test_assert_id(self, key):
         with pytest.raises(IntegrityError):
             key.assert_id(id, plaintext_changed)
 
+    def test_authenticated_encrypt(self, monkeypatch):
+        monkeypatch.setenv('BORG_PASSPHRASE', 'test')
+        key = AuthenticatedKey.create(self.MockRepository(), self.MockArgs())
+        plaintext = Chunk(b'123456789')
+        encrypted = key.encrypt(plaintext)
+        assert encrypted.endswith(plaintext.data)
+
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/key.py]: rather use authenticated = ... here as it is not encrypted?

File: src/borg/crypto.pyx [hunk 2]
```diff
@@ -6,6 +6,15 @@ from cpython.buffer cimport PyBUF_SIMPLE, PyObject_GetBuffer, PyBuffer_Release
 API_VERSION = 3
 
 
+cdef extern from "blake2/blake2b-ref.c":
+    ctypedef struct blake2b_state:
+        pass
+
+    int blake2b_init(blake2b_state *S, size_t outlen) nogil
+    int blake2b_update(blake2b_state *S, const void *input, size_t inlen) nogil
+    int blake2b_final(blake2b_state *S, void *out, size_t outlen) nogil
```
Comment (ThomasWaldmann) [inline src/borg/crypto.pyx]: alternatively, openssl 1.1 compatibility?

File: src/borg/archiver.py [hunk 3]
```diff
@@ -959,7 +959,7 @@ def _info_repository(self, repository, key, cache):
         else:
             encrypted = 'Yes (%s)' % key.NAME
         print('Encrypted: %s' % encrypted)
-        if key.NAME == 'key file':
+        if 'key file' in key.NAME:
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: key.NAME.startswith('key file') ?

File: src/borg/crypto.pyx (hunk 2 above)
Comment (enkore) [inline src/borg/crypto.pyx]: Yes, a compatible implementation can be written with anything that gives BLAKE2b, no special features (keying, tree hashing, personalization etc.) needed. This makes it compatible with the OpenSSL API (L. 237). Python 3.6 hashlib has the full API and is also compatible


However, to be very sure about that, I'll produce a small test program against OpenSSL 1.1

**Comment by enkore** [comment]
> However, to be very sure about that, I'll produce a small test program against OpenSSL 1.1

btdt, turns out OpenSSL 1.1 only supports specifically BLAKE2b-512 and BLAKE2s-256.

Since Python 3.6 on the other hand has the complete BLAKE2 API, including "digest_size".

Two options:

1. Use BLAKE2b-512 and truncate to 256 bits, to retain the option to change this code to OpenSSL 1.1
2. Leave as is. OpenSSL only becomes usable as they update their API: https://github.com/openssl/openssl/issues/980

**Comment by enkore** [comment]
CPython 3.6 works

```
Input:      037fb9b75b20d623f1d5a568050fccde4a1b7c5f5047432925e941a17c7a2d0d7061796c6f6164
Expected:   a22d4fc81bb61c3846c334a09eaf28d22dd7df08c9a7a41e713ef28d80eebd45
Calculated: a22d4fc81bb61c3846c334a09eaf28d22dd7df08c9a7a41e713ef28d80eebd45
Identical:  True

Input:      616263
Expected:   bddd813c634239723171ef3fee98579b94964e3bb1cb3e427262c8c068d52319
Calculated: bddd813c634239723171ef3fee98579b94964e3bb1cb3e427262c8c068d52319
Identical:  True

Input:      e944973af2256d4d670c12dd75304c319f58f4e40df6fb18ef996cb47e06367631323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637383930
Expected:   97ede832378531dd0f4c668685d166e797da27b47d8cd441e885b60abd5e0cb2
Calculated: 97ede832378531dd0f4c668685d166e797da27b47d8cd441e885b60abd5e0cb2
Identical:  True
```

**Comment by ThomasWaldmann** [comment]
What we also have to consider: linux (and other) package maintainers dislike it if one bundles 3rd party code.

So: is there something like a widespread libblake2b package we could use alternatively?

**Comment by enkore** [comment]
I didn't find any. There is libb2 but it's not commonly packaged (Debian: testing, Arch: No, Fedora: No, ...). It seems plausible that this is also why other software including CPython bundles the references implementation instead of using a library; there just isn't any.

**Comment by enkore** [comment]
I looked at whether the `-march` compiler option might ruin performance for generic distribution builds, which is not the case (somewhat surprisingly the performance of the C implementation does not stem from compiler optimizations).

**Comment by ThomasWaldmann** [comment]
http://packages.ubuntu.com/xenial/libdevel/libb2-dev it's in xenial.
https://github.com/BLAKE2/libb2/issues/5

So, I think we should alternatively link to that, so packagers who feel uncomfortable about bundled stuff can rip it out and link to their libb2 package (which they will have to create in case it is not there yet).

**Comment by enkore** [comment]
Done. Works, and the tests pass, too.

File: src/borg/blake2/libselect.h [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#ifdef BORG_USE_LIBB2
+#include <blake2.h>
+#else
+#include "blake2b-ref.c"
+#endif
```
Comment (ThomasWaldmann) [inline src/borg/blake2/libselect.h]: add a \n at eof.

Comment (ThomasWaldmann) [inline src/borg/blake2/libselect.h]: hmm, as this is rather part of borg than of blake2 ref implementation, move it to outside of this dir, so maintainers can just kill the whole blake2/ directory to remove it from the package if they use libb2?

Comment (enkore) [inline src/borg/blake2/libselect.h]: uh okay

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/pull/1819/commits/4074ea0fd779935f738355debb23659b4826c3d1 that looks strange on github.

**Comment by enkore** [comment]
Accidentally committed empty file; removed in fixup.

**Comment by ThomasWaldmann** [comment]
looks good. collapse some, merge.

**Pull request merged**